### ISSUE TITLE
Improve CI Caching

### ISSUE THREAD

**Issue by aluzzardi** [dagger/dagger#2140]
Generally speaking, buildkit cache export/import is pretty limiting because it's not "shared": every export cache overwrites the previous one, therefore CI runs are overwriting each other's cache. Persistent cache dirs are not part of the cache. And caching outside of GHA requires setting up some infrastructure (e.g. pointing to a registry).

This is a broad issue, keeping track of CI caching related issues:

- Using the GitHub Action, cache should work out of the box nicely
  - #1345
  - #2059
  - Ideally, without having to define a `scope` manually (can dagger or the GHA figure out a default scope?)
  - And without having to use `crazy-max/ghaction-github-runtime` (dagger GHA should do this? /cc @crazy-max)
- We need to find a path forward to support other CIs

NOTE: The above issues are pretty narrow in scope. However, taking a step back, we could think of something more drastic to overhaul the caching sub-system.

/cc @sipsma

**Comment by sipsma** [comment]
I have *A LOT* of thoughts here, many of which are intertwined given the connectedness of the problems. Starting for now with a brain dump just to get them out and see what thoughts others have. This will in time require a more organized design process given both its complexity and that it probably involves lots of collaboration with Buildkit.

---

Supporting other CIs:
1. **Pluggable cache backends** - One idea I brought up with @aluzzardi previously was to make cache export pluggable by adding support to Buildkit for exporting/importing remote cache using user-specified container images that serve an API to the buildkit daemon. This is essentially the same concept as LLB frontends but applied to the cache backend.
   1. This offers a lot of flexibility and potentially close-to-best-case performance. Its main downside is it will take a while to design and implement.
2. **Using local or registry exports as a proxy** - [The idea here](https://github.com/dagger/dagger/discussions/1875) to use either local or registry cache exports as the universal mechanism for plugging into some other cache backend.
   1. E.g. a local dir can be marked as a [GHA cache dir](https://github.com/actions/cache) (and I'm sure similar concepts exist in other CI systems). This, as noted in the discussion, has performance problems currently. Even if we improve those, this solution inherently has the overhead of putting an extra hop (local dir) between Buildkit and the actual backend where data is actually stored. TBD at what scale that overhead becomes a true bottleneck. Use of a network filesystem partially addresses this, but those add lots of complexity and are likely not available in every CI system.
   2. The idea of using a registry is intriguing too. In theory, you could deploy a registry local to the buildkit daemon (like a sidecar container), but the registry is some implementation that just proxies everything to another backend. I think I've heard of such things before, but have not investigated yet.

---

Persisting cache mounts:

To start with, Tonis's comment [here](https://github.com/docker/buildx/issues/244#issuecomment-602750160) is worth considering. Basically, most of the use cases for cache mounts are with downloaded package artifacts. In that case, is exporting them to a remote place and then re-downloading them actually faster than just re-downloading them from their original source (often a CDN, which is essentially a network cache)? It obviously depends on the use case, sometimes yes sometimes no. One compelling vote for yes in a particular case [is here](https://github.com/moby/buildkit/issues/1512#issuecomment-818856044).

1. **Explicitly converting cache mounts contents to layers** - The workaround Tonis suggests in the comment linked above (adding build steps that explicitly tar and untar the cache contents, essentially converting them to layers) might unblock this case for Dagger. We are an abstraction above Buildkit, so we can potentially hide the details of packing/unpacking the cache from users.
   1. This could provide a relatively quick solution to the problem. Main downside is that it has bad performance in many cases due to duplication of data, sometimes to the point where the overhead exceeds the performance benefits of the cache. There may be room for Buildkit to optimize this though, worth thinking about more.
1. **The [proposed mechanism](https://github.com/moby/buildkit/issues/1512) for providing drivers that back cache mounts** -  The examples there are drivers for a "docker volume" and a "bind mount" driver, but in theory more could be added.
   1. The "driver model" is intriguingly similar to the proposal for pluggable cache backends above, though the interfaces are so different (cache backend imports/exports layer tarballs, cache mounts just provides filesystem mount points). It's also similar to the "use local export as a proxy" option mention above for supporting different CIs, e.g. we could use the "bind mount" driver to map the Buildkit cache mount to a GHA cache dir. These similarities make me wonder if there's a better way of unifying these problems.
   2. One alternative idea might be to have a driver model that supports conversion of cache mount filesystems to layers. This allows you to plug cache mounts into the remote cache systems in a modular way. There would be a "dumb" driver that just converts the whole directory to a layer, but also smarter drivers that optimize layers for deduplication, or even use specialized knowledge of the layout (i.e. go cache dir, npm package dir, etc.) to create layers efficiently. This changes the interface enough that it could even just be combined with the idea of cache backends described above.
1. **Merging Package DAGs** - Another approach would be to solve the use case for combining package artifacts via MergeOp instead of cache mounts. Namely, MergeOp allows you to combine filesystem states together highly efficiently (i.e. hardlink trees). Instead of dumping package artifacts into a cache mount, builds can convert the package DAG into an LLB state per package, which then get combined together using MergeOp.
   1. The advantage here is that this integrates with remote cache export and Buildkit's caching mechanisms in general seamlessly, so our solutions to the other problems (e.g. of supporting different CI backends) benefit this use case too. It also works better for use cases where parts of the cache need to be included in final "runtime" images (e.g. interpreted languages like js or python where you download packages and then include them at runtime too, shared object libraries, etc.)
   1. The major downside is that this would need to be implemented per package system. I.e. you'd write a frontend that can convert an npm package.json to an LLB DAG that uses mergeop to combine everything together, but then you'd have to write another one for each other package manager.
   3. Additionally, while I've only ever seen cache mounts used to store package artifacts, in theory the concept is more general and some use cases for it probably don't translate to mergeop

---

Additive caches (this problem I've thought the least about so far, needs fleshing out)
1. [The current model](https://github.com/moby/buildkit/blob/master/cache/remotecache/v1/doc.go) used by Buildkit to export cache is not that far away from supporting this. I started working on Buildkit long after all this was added so I don't know the reasoning of the design well, but I have a vague memory that the main reason additive caching hadn't been done initially was just that there would be no Buildkit mechanism to manage the cache and it would grow forever. Need to confirm with Tonis though.
   1. If that is the only issue, it's definitely something that we could address. It might also fit in well the idea of pluggable cache backends; we could add cache management features (like querying what's there, pruning, etc.) to the cache backend API.
3. I think the solution in #2120 is promising too, but I haven't had a chance yet to investigate more on how it can be generalized

---

I'll just add as a final note that another approach that potentially solves all of these problems at once is to find a way to make the Buildkit worker cache distributed (basically, share snapshotter and content store state at the lowest level rather than share cache via exports on the top level). There is some interesting preliminary work in this space involving distributed containerd content stores, but this would require much much more investigation.

**Comment by sipsma** [comment]
One more addendum thought on the problem of wanting to share cache between CI runs. A relatively quick option that would require little to no changes to Buildkit would be to just cache the entirety of its state (i.e. `/var/lib/buildkit`) after the build is finished and Buildkit has shutdown. This obviously only works when Buildkit is being used by a single build at a time, but I think that's true of many CI cases (please correct me if wrong).

One issue here is that you might end up with a lot of duplication between packed data in the content store and unpacked snapshots. I suspect it could be fairly straightforward to add support to Buildkit's prune command to address this and allow you to trim the cached state to what you want before it shuts down though.

This is not exactly beautiful but it might be worth consideration as a stopgap given its simplicity relative to many other options.

**Comment by aluzzardi** [comment]
Thank you @sipsma, really appreciate all the context!

There's a lot to unpack here, added some ideas/feedback

### Supporting other CIs

#### Pluggable cache backends


> Pluggable cache backends - One idea I brought up with @aluzzardi previously was to make cache export pluggable by adding support to Buildkit for exporting/importing remote cache using user-specified container images that serve an API to the buildkit daemon. This is essentially the same concept as LLB frontends but applied to the cache backend.

I love this idea. Not sure how big of an effort it is (I suspect quite a bit)

#### Local cache

> Using local or registry exports as a proxy - https://github.com/dagger/dagger/discussions/1875 to use either local or registry cache exports as the universal mechanism for plugging into some other cache backend.

Did some testing here https://github.com/dagger/dagger/discussions/1875#discussioncomment-2445460, and it appears there's a huge overhead in using local exports.

I'm guessing it's because with registry, only the manifest plus modified layers get pushed, whereas with local it's a full blown export each time.

> E.g. a local dir can be marked as a [GHA cache dir](https://github.com/actions/cache)

Another downside is in addition to paying the local export overhead, we'd pay the GHA cache dir import/export overhead every time.

Even if nothing changed (or just a couple of layers), we'd have to download every single blob for buildkit to say "cached"

Compared to registry which only requires downloading the manifest and then lazily pulling in only inputs for the changed nodes (at least, that's how I assume it works)

To put it in perspective, if we take our own CI plan, if you change something irrelevant (e.g. `README.md`), we'd be pulling 800MB of local cache, running `dagger do`, nothing happens, exporting back the cache (~30 seconds), and then uploading it back to GHA cache.

#### Registry Backend

> The idea of using a registry is intriguing too. In theory, you could deploy a registry local to the buildkit daemon (like a sidecar container), but the registry is some implementation that just proxies everything to another backend.

Yeah that's definitely intriguing. It's more or less an alternative to the "pluggable cache backends" idea above, except instead of implementing it at the buildkit level, we'd implement it at the registry level.

### Persisting cache mounts

> To start with, Tonis's comment https://github.com/docker/buildx/issues/244#issuecomment-602750160 is worth considering. [...] In that case, is exporting them to a remote place and then re-downloading them actually faster than just re-downloading them from their original source

That's an interesting point of view for sure. It's true that many times, it's probably cheaper to just re-download. And perhaps a solution would be to "restrict" (e.g. through documentation / best practices) cache mounts to dependency downloading only.

Sometimes though, we've been using cache mounts for more than downloading. For instance, our `go` package uses a cache mount for build caching (`$GOCACHE`) -- e.g.  if a single go file changed, do not recompile all packages and dependencies.

I'm not sure how we could achieve the same result without cache mounts since any change in the code would invalidate the layer containing the cache.

The tricky part is how to "decouple" the inputs. Taking go as an example again, for go mod caching we'd need to only use `go.mod` / `go.sum` as inputs for the layer (not to invalidate the cache unless those files changed). For go build cache I don't know.

I think we need to take the time to profile how we're actually using the different types of cache and what we need from them. yarn is another interesting example where there's the yarn cache (CDN like) and then `node_modules` (which depends on `package.json` and `yarn.lock` -- pretty expensive to re-generate).

Still need some time to think through this :)

### Buildkit State Caching

> A relatively quick option that would require little to no changes to Buildkit would be to just cache the entirety of its state (i.e. /var/lib/buildkit) after the build is finished and Buildkit has shutdown. This obviously only works when Buildkit is being used by a single build at a time, but I think that's true of many CI cases (please correct me if wrong).

Thought about that in the past, however I think there's two pretty big downsides:

- Requires to download the entire state before proceeding (similar to the "local dir" downside above)

- Concurrency & Merging: For instance on dagger/dagger, there's one buildkit instance running per PR and one for main
  - To make this work, I guess we'd have to have a unique state per PR which means every PR starts from scratch
  - Similar issue as #2059, except with `--cache-from` we can merge multiple "states" together (whereas we can't merge two `/var/lib/buildkit` together, I think?)

**Comment by TomChv** [comment]
Thank you for all those information! I'm actively working so may I can share my point of view

> The idea of using a registry is intriguing too. In theory, you could deploy a registry local to the buildkit daemon (like a sidecar container), but the registry is some implementation that just proxies everything to another backend. I think I've heard of such things before, but have not investigated yet.

That's definitely a good idea but I'm curious how we could manage it in a CI context:
- I'm not sure we can run a "sidecare registry" in GitHub action, so the user will need to provide his own so cache will not work out of the box.
- Should we take advantage of `GitHub package` to store cache? Or deploy our own registry infrastructure shared by everyone so it's always available even in local?

On the other side, I think using a registry is the best way to ensure availability and performances are far more better than with local storage (as Andrea explained)

> I think the solution in https://github.com/dagger/dagger/pull/2120 is promising too, but I haven't had a chance yet to investigate more on how it can be generalized

It's basically the same as your merging idea but with a lot of limitation because it's done manually in the CI.
I started to investigate and the problem is : `how being CI agnostic?`

With GitHub we can use PR number, but what about GitLab CI or Tekton?
That's why I fall back in the registry option, that could be agnostic and the manifest take care of referencing layers. 

> A relatively quick option that would require little to no changes to Buildkit would be to just cache the entirety of its state (i.e. /var/lib/buildkit) after the build is finished and Buildkit has shutdown. This obviously only works when Buildkit is being used by a single build at a time, but I think that's true of many CI cases (please correct me if wrong).

I thought it was a good solution when I start working on cache, but it's too slow and not scalable. 
In my opinion, it would be worth to host your own Buildkit engine and share it in the CI  (like @gerhard did) than that.
And Andrea is right, it would be painful to merge `/var/lib/buildkit`, worth to create our own system.

But if I understand well, the main issue there is : https://github.com/dagger/dagger/issues/1345

Even with a registry or a new solution, since this issue exist, we will not be able to persist cache mount so all installation will not be cached and it will invalidate further steps? (Need a confirmation)

Something I do not understand is that in local with a simple dagger plan to build a go project : `#CacheDir` operation are exported in cache.

#### No cache

```
# With a clean buildkit daemon
dagger do build --log-level debug --log-format plain  
0.92s user 0.42s system 5% cpu 24.393 total
```

Build log

```
12:30PM DBG actions.build.container._exec | 
12:30PM DBG actions.build.container._exec | #5 Exec ["go", "build", "-v", "-tags", "", "-ldflags", "", "-o", "/output/", "."]
12:30PM DBG actions.build.container._exec | #5 sha256:d7f9b60bf47683f11a912d4cee73b82c168d5af48eadf07ac345fe721f10eb4c
12:30PM INF actions.build.container._exec | #5 0.061 go: downloading github.com/...
12:30PM DBG actions.build.container._exec | #5 DONE 2.3s
```

#### Import local cache

```shell
# With a clean buildkit daemon
time dagger do build --cache-to type=local,mode=max,dest=storage --cache-from type=local,src=storage
...
0.89s user 0.39s system 13% cpu 9.364 total
```

Build log

```shell
12:26PM DBG actions.build.container._exec | 
12:26PM DBG actions.build.container._exec | #6 Exec ["go", "build", "-v", "-tags", "", "-ldflags", "", "-o", "/output/", "."]
12:26PM DBG actions.build.container._exec | #6 sha256:2efc91a4ce7f06f98667d788cff651de31741e412025c143526698d295c9d145
12:26PM DBG actions.build.container._exec | #6 CACHED
```

**Why it's cached since `go.#Build` use `core.#CacheDir`**, it should not be saved according to https://github.com/dagger/dagger/issues/1345?

**Comment by TomChv** [comment]
Thank you for all those information! I'm actively working so may I can share my point of view

> The idea of using a registry is intriguing too. In theory, you could deploy a registry local to the buildkit daemon (like a sidecar container), but the registry is some implementation that just proxies everything to another backend. I think I've heard of such things before, but have not investigated yet.

That's definitely a good idea but I'm curious how we could manage it in a CI context:
- I'm not sure we can run a "sidecare registry" in GitHub action, so the user will need to provide his own so cache will not work out of the box.
- Should we take advantage of `GitHub package` to store cache? Or deploy our own registry infrastructure shared by everyone so it's always available even in local?

On the other side, I think using a registry is the best way to ensure availability and performances are far more better than with local storage (as Andrea explained)

> I think the solution in https://github.com/dagger/dagger/pull/2120 is promising too, but I haven't had a chance yet to investigate more on how it can be generalized

It's basically the same as your merging idea but with a lot of limitation because it's done manually in the CI.
I started to investigate and the problem is : `how being CI agnostic?`

With GitHub we can use PR number, but what about GitLab CI or Tekton?
That's why I fall back in the registry option, that could be agnostic and the manifest take care of referencing layers. 

> A relatively quick option that would require little to no changes to Buildkit would be to just cache the entirety of its state (i.e. /var/lib/buildkit) after the build is finished and Buildkit has shutdown. This obviously only works when Buildkit is being used by a single build at a time, but I think that's true of many CI cases (please correct me if wrong).

I thought it was a good solution when I start working on cache, but it's too slow and not scalable. 
In my opinion, it would be worth to host your own Buildkit engine and share it in the CI  (like @gerhard did) than that.
And Andrea is right, it would be painful to merge `/var/lib/buildkit`, worth to create our own system.

But if I understand well, the main issue there is : https://github.com/dagger/dagger/issues/1345

Even with a registry or a new solution, since this issue exist, we will not be able to persist cache mount so all installation will not be cached and it will invalidate further steps? (Need a confirmation)

Something I do not understand is that in local with a simple dagger plan to build a go project : `#CacheDir` operation are exported in cache.

#### No cache

```
# With a clean buildkit daemon
dagger do build --log-level debug --log-format plain  
0.92s user 0.42s system 5% cpu 24.393 total
```

Build log

```
12:30PM DBG actions.build.container._exec | 
12:30PM DBG actions.build.container._exec | #5 Exec ["go", "build", "-v", "-tags", "", "-ldflags", "", "-o", "/output/", "."]
12:30PM DBG actions.build.container._exec | #5 sha256:d7f9b60bf47683f11a912d4cee73b82c168d5af48eadf07ac345fe721f10eb4c
12:30PM INF actions.build.container._exec | #5 0.061 go: downloading github.com/...
12:30PM DBG actions.build.container._exec | #5 DONE 2.3s
```

#### Import local cache

```shell
# With a clean buildkit daemon
time dagger do build --cache-to type=local,mode=max,dest=storage --cache-from type=local,src=storage
...
0.89s user 0.39s system 13% cpu 9.364 total
```

Build log

```shell
12:26PM DBG actions.build.container._exec | 
12:26PM DBG actions.build.container._exec | #6 Exec ["go", "build", "-v", "-tags", "", "-ldflags", "", "-o", "/output/", "."]
12:26PM DBG actions.build.container._exec | #6 sha256:2efc91a4ce7f06f98667d788cff651de31741e412025c143526698d295c9d145
12:26PM DBG actions.build.container._exec | #6 CACHED
```

**Why it's cached since `go.#Build` use `core.#CacheDir`**, it should not be saved according to https://github.com/dagger/dagger/issues/1345?

**Comment by sipsma** [comment]
I got back to looking into this more tonight. Started looking through our CI in depth to get a handle on everything we're doing, started playing around with runs locally and in CI to see what might make sense to cache and what not, etc.

My main thought is that there's just a lot of areas where it's not clear whether remotely caching data will provide benefit or not, and the only way to get answers will be to run experiments and gather data. This is going to require adding tooling for summarizing timing information (i.e. how long did this vertex take to run vs. how long did it take to upload and download its cached result) and for orchestrating complex workflows (i.e. running multiple buildkit daemons, exchanging cache between them).
   * The risk is obviously that making tooling like this can become its own rabbit hole, so we'll need to build it iteratively, biased towards quick value. I have an idea on a relatively easy path to orchestrating complex workflows [here](https://github.com/dagger/dagger/issues/2246). Timing information I haven't dived into yet, but given Buildkit's support for tracing I'm sure we can find some open source tools to help us out here.

My immediate next steps are:
1. Continue getting a handle on everything our CI does and catalogue all the different things that could be remotely cached. I started tonight with just the `Build` workflow (and the associated actions), which is a good start but ultimately very simple (though still raises plenty of questions around go mod cache and go build cache). I suspect the integration tests will have a lot more to look into.
   1. Along the way, if I see any opportunities for caching that are just obviously beneficial and straightforward, I'll implement them. 
1. Look into tooling on gathering timing data and other information we'll need to start making actual decisions about what caching is worth improving. I just want something 1 step better than my current approach of running the build in plain progress mode and scrolling through the output to find how long each step ran :-). I'll start with tools built around tracing; I'm very open to suggestions here as I don't have a ton of prior experience with them.

---

Getting back on some comments:

> Concurrency & Merging: For instance on dagger/dagger, there's one buildkit instance running per PR and one for main

Ah okay, yeah the use case is different than I was thinking. Any sort of merging is out of the question for this approach so we can forget it.

> Sometimes though, we've been using cache mounts for more than downloading. For instance, our go package uses a cache mount for build caching ($GOCACHE) -- e.g. if a single go file changed, do not recompile all packages and dependencies.

Yeah these are the types of use cases that I think have the most potential for being beneficial when it comes to using remote caches: things that weren't just downloaded in the first place and have a very high ratio of `time to create` to  `compressed file size` (i.e. binaries, shared/static libs, etc.).

I've been refining my thoughts here a bit. There's two ways of approaching the caching of these artifacts. Let's say you are trying to build some binary by running `make foo`.

1. **Cache Dir** -  Just run `make foo` as an ExecOp with a Buildkit cache dir setup to hold all intermediate artifacts. This cache dir would be shared between every build and would be exported remotely for reimport by other builders (and CI runs).
1. **DAG-ify** - Break up the opaque `make foo` target into separate LLB vertices for each intermediate build step. I.e. in the Makefile it may turn out that `make foo` requires building 3 static libraries first before building the `foo` binary. Instead of implicitly doing that all in a single ExecOp, you could run 3 ExecOps for building the static libraries first and then combine those artifacts together and run the build for the `make foo` binary.

There are a lot of pros and cons to each approach. I started writing them out but it became a wall of text so I'll spare people now until my thoughts are more well formed and I have any data to back up my assertions. The gist is that the cache dir approach is simpler to setup but harder to manage, and the dag-ify approach is harder to setup but a much more robust solution once done.

> I'm not sure we can run a "sidecare registry" in GitHub action, so the user will need to provide his own so cache will not work out of the box.

The idea here was to just use the registry API as a universal API that can be proxied into any number of backends, so we could run that implementation using the container engine that Buildkit is running in, as a container in Buildkit itself using the new `AsyncExec` I have a PR out for, or probably even in the dagger client binary itself. Not that I guarantee any of this is a remotely good idea, that was just the thought :-) 

> Something I do not understand is that in local with a simple dagger plan to build a go project : #CacheDir operation are exported in cache.

What's happening there is that the result of the ExecOp has been cached, not the cache dir. When you imported the cache, Buildkit saw that it already had the result of the ExecOp and thus didn't not to run anything. If anything in the ExecOp definition (or one of its dependencies in the LLB DAG) had changed, then you wouldn't see the same cachedir contents as before (unless they were still in Buildkit's internal cache from a previous run).

**Comment by sipsma** [comment]
Had some promising ideas about the problems here which have resulted in more concrete plans. I like these ideas because they don't require any upstream Buildkit changes, which makes it faster to get experimenting and potentially get useful features if the ideas pan out.

---

**Cache Dir Export**
New idea: run an`AsyncExec` with the shared cache dir mounted, put a daemon in there that handles the syncing of the cache dir with an external backend. For example, if we want to sync a cache mount with GHA, we'd start a daemon in an `AsyncExec` that watches the dir and syncs it with the cache backend via the GHA API. We can start the daemon at the very beginning of the build to start populating the cache and later block the dagger client on it flushing everything before exiting.

The best part of this is that, unlike my previous ideas, it should already work today with no upstream changes.

The second best part of this is the modularity of being able to easily add different implementations for different backends. We can have implementations not only for other CI systems and storage backends, but even stuff like continuous syncing with a dir local to the client (using mutagen for example) which unlocks some interesting possibilities...

Only downside is that getting "continuous two-way syncing" right can be much easier said than done, but I think we can start rudimentary, e.g. just synchronous one-way push/pull at the beginning and end of a build session, may not even need `AsyncExec`.

---

**Scopes+Additive Caches on Any Backend**

Was thinking about ways of making this seamless+generic so that users can just get this caching working nicely with whatever backend they want and not having to construct complicated `cache-from` and `cache-to` arguments themselves (or use wrapper scripts we provide). My current favorite idea is to support setting `cache-from` and `cache-to` via cue, so users can import pre-baked, reusable solutions. Not sure yet, but I think this could be another top-level field under `Plan` (i.e. a sibling to `client` and `actions` called something like `cache`).

An example implementation is one that sets up GHA cache with scopes similar to #2120. But others would achieve the same end effect using different backends. As another example, here's a possible recipe for getting additive caching working a generic registry backend:
1. Builds can export cache to a repository using a tag unique to them (e.g. pr number) but grouped with other tags using prefixes (i.e. `dagger-ci-build-foobar`, with `-foobar` being the unique id).
2. Then, to setup cache imports we can query the registry for all tags (`ListTags` seems to be in the APIs I checked) and filter using prefixes to find the ones we want to import (very similar to how GHA cache restore keys work)
3. This will need some way of automatically evicting and/or manually pruning cache so it doesn't grow forever
4. I will also need to do some research and experiments to see how buildkit does when importing a lot of cache sources, i.e. what's the scale at which it slows down and how easy/hard is it to fix.

The idea is that users would just specify something high-level in their cue like what registry/repo to use and what unique ID to use, then all the cache import/export settings would get figured out for them.

Finally, there is still the problem of how to get this working on everything else people might want to export cache to besides GHA and a registry. I think the idea of using the registry API as a proxy to other backends may be the best way to start. It's ugly relative to the options that require upstream Buildkit changes, but it's possible to get going quickly (just run the proxy implementation via an AsyncExec) and can serve as a testing grounds for what we might want to upstream more cleanly someday.

---

**TODOs**
Combining these ideas with the next steps I mentioned in my previous comment, ordered roughly by my current priority:
1. Test idea of setting up cache-from and cache-to using cue (instead of command line flags/env vars)
   * I think this has the most potential to be valuable to users quickest
1. Test idea of syncing a cache dir with GHA using `AsyncExec`
   * It remains to be seen what cases this is actually beneficial in as discussed previously, but having this setup will at least let us start experimenting
1. Hook things up to a tracing API for getting timing metrics
1. Test idea of additive registry caches
1. Test idea of registry api as proxy to other backends

All of these tests will just start as basic proof-of-concepts, after which we can assess if they are promising on their own or if they indicate that we need to do something more time-consuming like upstream features to Buildkit.

Throughout all this I'm going to continue playing with our existing CI and make small improvements here and there too.

**Comment by aluzzardi** [comment]
Sounds like a solid plan!

- **Scopes+Additive Caches on Any Backend**

`cache-from` / `cache-to` using CUE is a very interesting idea. I'm wondering how "involved" the user would need to be to set this up (e.g. would this require setting up some CI specific settings on the Plan such as repository etc).

Alternatively, we could split that logic between the CI specific integration code (e.g. GH Action, Circle CI ORB) and dagger itself.

Basically, the GH action code has all the data we need (repository, PR number, remote ref, etc). And dagger has the other "sub scope": action  (e.g. `do <ACTION NAME>`) and which project we're running (e.g. `-p <file.cue>`)

In theory, combining those two, we have everything we need to do scoping/additive zero-config caching (end up with something like `dagger-cache-[PR / HEAD ID]-[PROJECT]-[ACTION]`). And from the action we also know the target branch of the PR (to import cache).

But I do see the appeal of doing that in CUE -- we could implement those "drivers" as regular dagger packages rather than custom integrations like the action, definitely worth exploring.

Another TODO regarding the GHA: no matter the approach, I think we should "merge" the logic of `crazy-max/ghaction-github-runtime@v1` inside the dagger GHA. That action exports some caching information to buildkit and right now users have to use both actions to get caching to work. I think it would be better if everything worked out of the box when using the dagger GHA.

- **Cache Dir Export**

Yep that sounds great -- no upstream changes, we can experiment right away. I like that :)

**Comment by sipsma** [comment]
> Alternatively, we could split that logic between the CI specific integration code (e.g. GH Action, Circle CI ORB) and dagger itself.

True, both sides of this will need to be involved for sure, I guess it's just about where we draw the border exactly. My hope is that putting more of the logic in dagger packages would result in more generic and re-usable code across multiple integrations. Then the CI-specific code code just be used to setup env vars and other client state that can be read by the dagger package as args essentially (just using the existing `client` mechanisms).

Tough to say if this theory will work out in practice though, could just lead to more code if it ends up being hard to create re-usable code. I'll just keep an open mind about this when I start trying it out.

> Another TODO regarding the GHA: no matter the approach, I think we should "merge" the logic of crazy-max/ghaction-github-runtime@v1 inside the dagger GHA. 

Right, that should be an easy win too. I'll take a look at it.

**Comment by TomChv** [comment]
> Test idea of setting up cache-from and cache-to using cue (instead of command line flags/env vars)
I think this has the most potential to be valuable to users quickest

I think we could use the Client API to let user specify where store cache.

Here's a quick example from my mind

```
dagger.#Plan & {
  client: cache: "my-cool-plan": {
     type: "registry"
     ref: "localhost:6000/toto"
     auth: {...} // How should we manage authentication?
  }
}
```

And so this will be transform in `--cache-to type=regsitry,ref=localhost:6000/toto:my-cool-plan`

WDYT?

> New idea: run anAsyncExec with the shared cache dir mounted, put a daemon in there that handles the syncing of the cache dir with an external backend

I really like that idea but may with time it will be complex to maintain those daemons, indeed, we should also find a way to prune.

**Comment by sipsma** [comment]
> I really like that idea but may with time it will be complex to maintain those daemons, indeed, we should also find a way to prune.

Agreed but I also can't think any way to avoid something that needs to be maintained. Even if we, say, upstream support for a particular backend to Buildkit (similar to how Buildkit supports GHA today), we will still need to maintain the code there. From that perspective, I'd actually rather maintain something more modular that can be upgraded on the fly without having to upgrade the entirety of Buildkit. This probably also makes it easier for external contributors to implement their own functionality.

> WDYT?

Yeah I like it, this is along the lines of what I was imagining, couple minor thoughts:
1. Rather than using a `type: string` to switch between different implementations, I'd be interested in if we can use separate cue definitions that just output a common interface of `from` and `to` values
1. I wonder if it might make more sense to have `cache` be its own field on the same level as `client` and `actions` rather than under `client`. Just because `cache` doesn't necessarily involve resources local to the `client`. This is mostly a superficial issue though.

So, just iterating on your suggestion, maybe something like:
```
dagger.#Plan & {
  client: { ... }
  cache: {
    "my-registry-cache": #cache.Registry & {
        ref: "localhost:6000/toto"
        auth: { token: client.secret.token } // can read in stuff from the client
    } 
    "my-s3-cache": #cache.S3 & {
        bucket: "foobar-bucket"
        aws_secret_key: {...}
     }
  }
  actions: {...} // all actions are now backed by the cache and new results will be uploaded
}
```

Each entry under `cache` can output both `from` and/or `to` values that are then passed to the Buildkit solver as the cache import/export values (though for now we will be limited to a single cache output as you know). If a particular implementation also needs to setup some extra state (such as an s3 backend needing to setup a container that translates registry API calls to S3 ones), then we can support that too by allowing an output of type `daemon: #core.AsyncExec` or something maybe.

**Comment by sipsma** [comment]
I guess another nice thing about supporting cache settings like this inside dagger plans/actions is that you could potentially use dagger to setup the cache infra. I.e. I could provision a registry using Dagger (or just do a noop if it already exists) and then use that as the cache backend for the rest of my plan. This sounds like it could complicate our cue workflow, but probably something that can be done in the fullness of time if nothing else?

**Comment by sipsma** [comment]
https://github.com/dagger/dagger/issues/2365 is related. Besides the immediate problems there, it also raises an interesting question in general: if the backing cache is sometimes fast, sometimes slow (due to throttling or anything else), how does a user determine if caching is actually beneficial?

Would it make sense to have the ability to simultaneously execute a vertex and import cache for it, basically letting the two race and choosing whichever is fastest? This **might** sometimes result in the most consistently fast behavior, but at the expense of resources (cpu time and network bandwidth). How would we guide users on whether they should do this or not?

Another possibility is to just allow the cache import to timeout (with a configurable value) and then fallback to executing the vertex, but how would we guide users to set an optimal timeout value (or automatically set it for them)?

I don't have any answers yet, just making sure these types of questions are tracked.

**Comment by sipsma** [comment]
> Should we fix it in Buildkit? It feels like a strong limitation for the future of dagger cache

Yes for sure. There is [an open PR](https://github.com/moby/buildkit/pull/2760) to add support for multiple non-cache exports. I think once that's merged it would make sense to add the same ability but for cache exports.

> It opens a lot of possibility, but we should leave this for now, let's focus on cache

Agreed, I'm thinking long-term here, not proposing we do everything right away :-)

**Comment by mircubed** [comment]
When Erik gets back from vacation, we can review this issue to see how we can break it down into smaller bitesize issues that can be tracked in an easier way.

**Comment by sipsma** [comment]
A lot of this is still relevant in an abstract way but we've moved entirely off the previous CI system and will continue to, so closing this as no longer actionable. Can be revisited for ideas on CI caching in the future.
